src/arakoon-cluster.c

```diff
@@ -110,35 +111,80 @@ void arakoon_cluster_free(ArakoonCluster *cluster) {
         arakoon_mem_free(cluster);
 }
 
+static size_t cluster_size(ArakoonCluster * const cluster) {
+        size_t n = 0;
+        ArakoonClusterNode *node = NULL;
+
+        for (node = cluster->nodes; node != NULL; node = _arakoon_cluster_node_get_next(node)) {
+                ++n;
+        }
+
+        return n;
+}
+
+static int initial_timeout(size_t num_nodes,
+                           const ArakoonClientCallOptions * const options) {
+        const int total_timeout = options != NULL ?
+                arakoon_client_call_options_get_timeout(options) :
+                ARAKOON_CLIENT_CALL_OPTIONS_DEFAULT_TIMEOUT;
+        int timeout = 0;
+
+        if (total_timeout > 0) {
+                /* Worst case (cf. arakoon_cluster_connect_master):
+                 * connection to all nodes but the last one times out, and afterwards we
+                 * still connect to and verify the master, so that's (num_nodes + 1) * 2
+                 */
+                num_nodes = (num_nodes + 1) * 2;
+                timeout = total_timeout / num_nodes + ((total_timeout % num_nodes == 0) ? 0 : 1);
+        } else {
+                timeout = total_timeout;
+        }
+
+        return timeout;
+}
+
+static void adjust_timeout(int per_node_timeout, int * timeout) {
+        assert(timeout);
+
+        if (*timeout > 0) {
+                assert(per_node_timeout > 0);
+                *timeout += per_node_timeout;
+        } else if (per_node_timeout >= 0) {
+                *timeout = per_node_timeout;
+        } else {
+                *timeout = ARAKOON_CLIENT_CALL_OPTIONS_INFINITE_TIMEOUT;
+        }
+}
+
 arakoon_rc arakoon_cluster_connect_master(ArakoonCluster * const cluster,
-    const ArakoonClientCallOptions * const options) {
+                                          const ArakoonClientCallOptions * const options) {
         ArakoonClusterNode *node = NULL;
         arakoon_rc rc = 0;
         char *master = NULL;
-        int timeout = ARAKOON_CLIENT_CALL_OPTIONS_DEFAULT_TIMEOUT;
+        size_t num_nodes = cluster_size(cluster);
+        const int per_node_timeout = initial_timeout(num_nodes,
+                                                     options);
+        int timeout = per_node_timeout;
 
         FUNCTION_ENTER(arakoon_cluster_connect_master);
 
         ASSERT_NON_NULL_RC(cluster);
 
         _arakoon_log_debug("Looking up master node");
 
-        timeout = options != NULL ?
-                arakoon_client_call_options_get_timeout(options) :
-                ARAKOON_CLIENT_CALL_OPTIONS_DEFAULT_TIMEOUT;
-
         /* Find a node to which we can connect */
         node = cluster->nodes;
         while(node != NULL) {
+                num_nodes -= 1;
                 rc = _arakoon_cluster_node_connect(node, &timeout);
+                adjust_timeout(per_node_timeout, &timeout);
```

Comment:
I'm not sure I'm reading this correctly.
Is every next node getting bigger and bigger timeout?

Reply:
The idea is indeed that the next operation can use any time left over from the previous call. So if connecting to the first node + asking for the master happens rather quickly, the final connect (+ verification that it is the master) to the master can use up the remaining time.

Reply:
num_nodes = 3;
config(timeout) = 3;
per_node_timeout = 1; 3 / 
timeout = 1;

all nodes are down, so all connects timeouts

1st loop iter
 wait(timeout) -> 1s
 adjust_timeout -> timeout = timeout + per_node_timeout = 2

2nd loop iter
 wait(timeout) -> 2s
 adjust_timeout -> timeout = timeout + per_node_timeout = 3

3rd loop iter
 wait(timeout) -> 3s

so, instead of waiting 3s we waited 6s.
am I missing something?

Reply:
The timeout is decremented by the `_arakoon_cluster_node_{connect,who_master}` calls (which is the source of the original bug, i.e. the first call could eat up the whole timeout). So the scenario you're describing would be handled like this:
* connect to first node with timeout of 0.5 secs -> expires
* connect to second node with timeout of 0.5 secs -> expires
* connect to third node with timeout of 0.5 secs -> expires
IOW it's actually returning before the timeout allocated by the caller (which might be worth revisiting).

Reply:
Got it. I like deadlines more than timeouts.